x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/action/TransportGetStackTracesAction.java

```diff
@@ -371,33 +403,49 @@ private void searchEventGroupedByStackTrace(
 
-                Terms executableNames = searchResponse.getAggregations().get("group_by");
-                for (Terms.Bucket executableBucket : executableNames.getBuckets()) {
-                    String executableName = executableBucket.getKeyAsString();
-
-                    Terms threads = executableBucket.getAggregations().get("group_by");
-                    for (Terms.Bucket threadBucket : threads.getBuckets()) {
-                        String threadName = threadBucket.getKeyAsString();
-
-                        Terms hosts = threadBucket.getAggregations().get("group_by");
-                        for (Terms.Bucket hostBucket : hosts.getBuckets()) {
-                            String hostID = hostBucket.getKeyAsString();
-
-                            Terms stacktraces = hostBucket.getAggregations().get("group_by");
-                            for (Terms.Bucket stacktraceBucket : stacktraces.getBuckets()) {
-                                Sum count = stacktraceBucket.getAggregations().get("count");
-                                int finalCount = resampler.adjustSampleCount((int) count.value());
-                                if (finalCount <= 0) {
-                                    continue;
+                for (Terms.Bucket samplingFrequencyBucket : samplingFrequencies.getBuckets()) {
+                    log.debug(
+                        "Using sampling frequency [{}] for [{}] stacktrace events.",
+                        samplingFrequencyBucket.getKeyAsString(),
+                        totalCount
+                    );
+                    final double samplingFrequency = samplingFrequencyBucket.getKeyAsNumber().doubleValue();
+                    final double samplingFactor = maxSamplingFrequency / samplingFrequency;
+
+                    Terms executableNames = samplingFrequencyBucket.getAggregations().get("group_by");
+                    for (Terms.Bucket executableBucket : executableNames.getBuckets()) {
+                        String executableName = executableBucket.getKeyAsString();
+
+                        Terms threads = executableBucket.getAggregations().get("group_by");
+                        for (Terms.Bucket threadBucket : threads.getBuckets()) {
+                            String threadName = threadBucket.getKeyAsString();
+
+                            Terms hosts = threadBucket.getAggregations().get("group_by");
+                            for (Terms.Bucket hostBucket : hosts.getBuckets()) {
+                                String hostID = hostBucket.getKeyAsString();
+
+                                Terms stacktraces = hostBucket.getAggregations().get("group_by");
+                                for (Terms.Bucket stacktraceBucket : stacktraces.getBuckets()) {
+                                    Sum count = stacktraceBucket.getAggregations().get("count");
+                                    int finalCount = resampler.adjustSampleCount((int) Math.round(count.value() * samplingFactor));
```

Comment:
Can we skip the resampling if `samplingFactor == 1.0` ?

Reply:
No we can't. The resampler takes care of stratified downsampling (in steps 5^N).